components/serverless/api/v1beta1/function_types.go

```diff
@@ -0,0 +1,206 @@
+	//TODO: we should check if it is possible
+	// +optional
+	CRDSource *CRDSource `json:"crdSource"`
+	// +optional
+	GitRepositorySource *GitRepositorySource `json:"gitRepositorySource"`
+	// +optional
+	InlineSource *InlineSource `json:"inlineSource"`
+}
+
+type InlineSource struct {
+	Source       string `json:"source"`
+	Dependencies string `json:"dependencies"`
+}
+
+type GitRepositorySource struct {
+	Name      string `json:"name"`
+	Namespace string `json:"namespace"`
+}
+
+type CRDSource struct {
+	Api        string `json:"api"`
+	Version    string `json:"version"`
+	Kind       string `json:"kind"`
+	Name       string `json:"name"`
+	Namespaces string `json:"namespaces"`
+	Path       string `json:"path"`
+	BaseDir    string `json:"baseDir"`
+	// +optional
+	Credentials *Credentials `json:"credentials"`
+}
+
+type ArtifactSource struct {
+	//TODO: maybe it's worth do distinguish between local and external, as we know the internal URL
+	URL     string `json:"url"`
+	BaseDir string `json:"baseDir"`
+	// +optional
+	Credentials *Credentials `json:"credentials"`
+}
+
+type ConfigMapSelector struct {
+	Name      string `json:"name"`
+	Namespace string `json:"namespace"`
+}
+
+//TODO: To discuss auth options
+type Credentials struct {
+	// +optional
+	BasicAuth *BasicAuth `json:"basicAuth"`
+	// +optional
+	Oauth *Oauth `json:"oauth"`
+	// +optional
+	JWTToken *v1.SecretKeySelector `json:"token"`
+	// +optional
+	PersonalAccessToekn *v1.SecretKeySelector `json:"personalAccessToekn"`
+}
+
+//TODO: should we enforce using secret or give the ability to use configmaps or plain values
+type BasicAuth struct {
+	User     v1.SecretKeySelector `json:"user"`
+	Password v1.SecretKeySelector `json:"password"`
+}
+
+type Oauth struct {
+	OauthURL string               `json:"oatuhURL"`
+	User     v1.SecretKeySelector `json:"user"`
+	Password v1.SecretKeySelector `json:"password"`
+}
+
+type Template struct {
+	// +optional
+	Labels map[string]string `json:"labels,omitempty"`
+	// +optional
+	Annotations map[string]string `json:"annotations"`
+}
+
+// FunctionSpec defines the desired state of Function
+type FunctionSpec struct {
+	// INSERT ADDITIONAL SPEC FIELDS - desired state of cluster
+	// Important: Run "make" to regenerate code after modifying this file
+
+	// Foo is an example field of Function. Edit function_types.go to remove/update
+	Runtime string `json:"runtime"`
+
+	//+optional
+	CustomRuntimeConfiguration *ConfigMapSelector `json:"customRuntimeConfiguration"`
+
+	Source Source `json:"source"`
+
+	// Env defines an array of key value pairs need to be used as env variable for a function
+	Env []v1.EnvVar `json:"env,omitempty"`
+
+	// +optional
+	RuntimeProfile *string `json:"runtimeProfile"`
+
+	// +optional
+	Resources *v1.ResourceRequirements `json:"resources,omitempty"`
+
+	// +optional
+	BuildProfile *string `json:"buildProfile"`
+
+	// +optional
+	BuildResources *v1.ResourceRequirements `json:"buildResources,omitempty"`
```

Comment:
How do these fields work together?

Reply:
I wanted to make it similar to the [k8s volume source](https://github.com/kubernetes/api/blob/master/core/v1/types.go#L49)
Generally you have to choose only one of them.

Reply:
You can choose predefined profile or create custom one.

Reply:
What about this:
```
build:
 - profile
 - resources
```
```
runtime:
 - profile
 - resources
```
And only one of those options should be filled.

Reply:
New idea
```
requirements:
 -build:
    -profile
    -resources
 -runtime
  - profile
  - resources
  ```

Reply:
I like the latest idea more. And you are right about the `runtime` name. I would also suggest `deployment` or `instance` for this.

Reply:
What do you think about `function`?
```
requirements/resourceConfig:
 -build:
    -profile
    -resources
 -function
  - profile
  - resources
  ```

Reply:
I've updated the cr.